core-rust/state-manager/src/metrics.rs

```diff
@@ -473,3 +486,73 @@ impl SelfProposalMissTracker {
         self.gauge.add(new_missed_count - outdated_missed_count);
     }
 }
+
+/// A number of most recent [`ProposerTimestampDatapoint`]s to track for metrics purposes.
+const PROGRESS_RATE_HISTORY_LEN: usize = 10;
+
+/// A `proposer_timestamp_ms` captured at a specific `wallclock_epoch_sec`.
+#[derive(Debug, Clone, Copy)]
+struct ProposerTimestampDatapoint {
+    wallclock_epoch_sec: f64,
+    proposer_timestamp_ms: i64,
+}
+
+impl ProposerTimestampDatapoint {
+    /// Captures the given `proposer_timestamp_ms` at the current wall-clock.
+    pub fn at_current_wallclock(proposer_timestamp_ms: i64) -> Self {
+        Self {
+            wallclock_epoch_sec: SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs_f64(),
+            proposer_timestamp_ms,
+        }
+    }
+
+    /// Calculates the rate of the `proposer_timestamp_ms` measured relative to the given reference
+    /// point.
+    pub fn proposer_timestamp_rate_since(&self, reference: &ProposerTimestampDatapoint) -> f64 {
+        let delta_proposer_timestamp_ms =
+            self.proposer_timestamp_ms - reference.proposer_timestamp_ms;
+        let delta_proposer_timestamp_sec = (delta_proposer_timestamp_ms as f64) / 1000.0;
+        let delta_wallclock_sec = self.wallclock_epoch_sec - reference.wallclock_epoch_sec;
+        delta_proposer_timestamp_sec / delta_wallclock_sec
+    }
+}
+
+/// A higher-level metric helper, tracking a recent rate of `proposer_timestamp_ms` committed to the
+/// ledger.
+pub struct ProposerTimestampProgressRateTracker {
+    buffer: Mutex<RingBuffer<ProposerTimestampDatapoint, PROGRESS_RATE_HISTORY_LEN>>,
+    gauge: Gauge,
+}
+
+impl ProposerTimestampProgressRateTracker {
+    /// Creates a new tracker and registers its resulting [`Gauge`] (with the given options) at the
+    /// given registry.
+    /// Note: the [`LockFactory`] is required to ensure a thread-safe access to a ring-buffer used
```

Comment:
yeah it would increase the chances (A LITTLE BIT) of not staying like that forever... What would its impl be? (most importantly: would it still require a `LockFactory` dep, or are you thinking about something smarter)

On the other hand, the metrics lib we use has all its metric primitives thread-safe. Our new metric primitive happens to be a bit more complex and requires Mutex (well technically in Java it wouldn't require one...) - is it wrong to keep it thread-safe then? (I am afraid that callers may assume it - especially after you mentioned the future in your comment).